src/encoded/types/workflow.py

```diff
@@ -135,6 +140,367 @@
     }
 }
 
+def get_unique_key_from_at_id(at_id):
+    if not at_id:
+        return None
+    at_id_parts = at_id.split('/')
+    return at_id_parts[2]
+
+DEFAULT_TRACING_OPTIONS = {
+    'max_depth_history' : 6,
+    'max_depth_future' : 6,
+    "group_similar_workflow_runs" : True,
+    "track_performance" : False
+}
+
+
+def trace_workflows(original_file_item_uuid, request, file_item_input_of_workflow_run_uuids, file_item_output_of_workflow_run_uuids, options=None):
+
+    if options is None:
+        options = DEFAULT_TRACING_OPTIONS
+
+    if options.get('track_performance'):
+        pr = cProfile.Profile()
+        pr.enable()
+
+    uuidCacheModels = {}
+    uuidCacheTracedHistory = {}
+    steps = [] # Our output
+
+
+    def get_model_by_uuid(uuid, key = None):
+        # TODO: Check for hasattr(model, 'source') and raise Exception? (after we have something setup to handle it)
+        model = None
+        cacheKey = uuid
+        if key is not None:
+            cacheKey = key + ':' + uuid
+
+        model = uuidCacheModels.get(cacheKey)
+        if model is not None:
+            return model
+
+        if key is None:
+            model = request.registry['connection'].storage.get_by_uuid(uuid)
+        else:
+            model = request.registry['connection'].storage.get_by_unique_key(key, uuid)
+        uuidCacheModels[cacheKey] = model
+        return model
+
+
+    def group_files_by_workflow_argument_name(set_of_files):
+        files_by_argument_name = OrderedDict()
+        for f in set_of_files:
+            arg_name = f.get('workflow_argument_name')
+            if arg_name:
+                if files_by_argument_name.get(arg_name) is None:
+                    files_by_argument_name[arg_name] = [f]
+                else:
+                    files_by_argument_name[arg_name].append(f)
+        return files_by_argument_name
+
+
+    def filter_workflow_runs(workflow_run_tuples):
+        if len(workflow_run_tuples) < 3:
+            return (workflow_run_tuples, [])
+        filtered_tuples = []
+        filtered_out_tuples = []
+        tuples_by_workflow = {}
+        for workflow_run_uuid, in_file_uuid, workflow_run_model_obj, in_file in workflow_run_tuples:
+            workflow_atid = workflow_run_model_obj.get('workflow')
+            if not tuples_by_workflow.get(workflow_atid):
+                tuples_by_workflow[workflow_atid] = []
+            tuples_by_workflow[workflow_atid].append((workflow_run_uuid, in_file_uuid, workflow_run_model_obj, in_file))
+
+        def get_date_created_from_tuple(wfr_tuple):
+            workflow_run_uuid, in_file_uuid, workflow_run_model_obj, in_file = wfr_tuple
```

Comment:
I think the same-name`filter_workflow_runs` variables aren't overwritten since the `for` block ends before the func definition and hopefully are de-referenced (or at least that was my intent lol). But ya either way don't have any actual need to unpack the tuple (other than to access the value). Will try to improve.